File: ios/CaboGame/Core/CaboGameEngine.swift
import Foundation

enum GameRuleError: LocalizedError {
    case notEnoughPlayers
    case invalidPlayer
    case invalidPhase
    case emptyDeck
    case noDiscardCard
    case invalidCardIndex
    case noPendingDraw
    case invalidSpecialCard
    case caboAlreadyCalled
    case roundFinished

    var errorDescription: String? {
        switch self {
        case .notEnoughPlayers: return "Need at least 2 players."
        case .invalidPlayer: return "It is not this player's turn."
        case .invalidPhase: return "This action is not valid right now."
        case .emptyDeck: return "Deck is empty."
        case .noDiscardCard: return "No card in discard pile."
        case .invalidCardIndex: return "Selected card index is invalid."
        case .noPendingDraw: return "No drawn card is pending."
        case .invalidSpecialCard: return "This card has no special power."
        case .caboAlreadyCalled: return "Cabo has already been called this round."
        case .roundFinished: return "Round already finished."
        }
    }
}

struct CaboGameEngine {
    /// Maximum length of a single player turn after the initial peek phase.
    /// Matches the Android engine's `TURN_TIMEOUT_SECONDS`.
    static let turnTimeoutSeconds: TimeInterval = 20

    private(set) var state: GameState

    init(players: [Player] = []) {
        self.state = GameState(players: players)
    }

    mutating func addPlayer(name: String) -> UUID {
        let newPlayer = Player(name: name)
        state.players.append(newPlayer)
        return newPlayer.id
    }

    mutating func load(state: GameState) {
        self.state = state
    }

    mutating func startGame() throws {
        guard state.players.count >= 2 else { throw GameRuleError.notEnoughPlayers }
        state.deck = .caboDeck()
        state.discardPile = []
        state.currentPlayerIndex = 0
        state.phase = .initialPeek
        state.pendingDraw = nil
        state.winnerID = nil
        state.caboCallerID = nil
        state.rematchRequestedByHost = false
        state.readyPlayerIDs = []
        state.hasStarted = true
        state.playersFinishedInitialPeek = 0
        state.initialPeekGraceEndsAt = Date().addingTimeInterval(15)
        state.currentTurnEndsAt = nil
        state.kingSwapHighlight = nil
        state.initialPeekedIndicesByPlayerIndex = Array(repeating: [], count: state.players.count)

        for idx in state.players.indices {
            state.players[idx].hand = Array(repeating: nil, count: 4)
            state.players[idx].roundsToSkip = 0
        }

        for slot in 0..<4 {
            for idx in state.players.indices {
                let card = try drawFromDeckInternal()
                state.players[idx].hand[slot] = card
            }
        }

        // Seed discard pile so "draw from discard" is available.
        let firstDiscard = try drawFromDeckInternal()
        state.discardPile.append(firstDiscard)
    }

    mutating func peekInitialCard(for playerID: UUID, at handIndex: Int) throws -> Card {
        guard state.phase == .initialPeek else { throw GameRuleError.invalidPhase }

        let playerIndex = try indexOfPlayer(playerID)
        guard state.players[playerIndex].hand.indices.contains(handIndex) else {
            throw GameRuleError.invalidCardIndex
        }
        guard !state.initialPeekedIndicesByPlayerIndex[playerIndex].contains(handIndex) else {
            throw GameRuleError.invalidCardIndex
        }

        state.initialPeekedIndicesByPlayerIndex[playerIndex].append(handIndex)
        guard let card = state.players[playerIndex].hand[handIndex] else { throw GameRuleError.invalidCardIndex }

        // Count completion when a player reaches 2 peeked cards.
        if state.initialPeekedIndicesByPlayerIndex[playerIndex].count == 2 {
            state.playersFinishedInitialPeek += 1
            // Do not start the game early just because everyone peeked.
            // The phase transition is governed solely by `initialPeekGraceEndsAt`.
        }

        return card
    }

    mutating func beginMainTurnsAfterInitialPeekIfReady(now: Date = Date()) {
        guard state.phase == .initialPeek else { return }
        guard let graceEndsAt = state.initialPeekGraceEndsAt else { return }
        guard now >= graceEndsAt else { return }

        state.phase = .waitingForDraw
        state.currentPlayerIndex = 0
        state.initialPeekedIndicesByPlayerIndex = Array(repeating: [], count: state.players.count)
        state.playersFinishedInitialPeek = 0
        state.initialPeekGraceEndsAt = nil
        state.kingSwapHighlight = nil
        startTurnTimer(now: now)
    }

    /// Forces the current player's turn to end because the per-turn timer
    /// expired. Behavior depends on the phase the player got stuck in:
    ///  - `waitingForDraw`            : skip the turn outright.
    ///  - `waitingForPlacementOrDiscard`: discard the pending drawn card without
    ///                                    applying its effect, then advance.
    ///  - `waitingForSpecialResolution`: forfeit the J/Q/K effect (the card is
    ///                                    already on the discard pile), advance.
    /// Returns `true` if the turn was actually ended, `false` if there was
    /// nothing to do (e.g. game over or initial peek phase).
    @discardableResult
    mutating func enforceTurnTimeoutIfNeeded(now: Date = Date()) -> Bool {
        guard state.winnerID == nil else { return false }
        guard state.phase != .initialPeek else { return false }
        guard let deadline = state.currentTurnEndsAt, now >= deadline else { return false }

        if state.phase == .waitingForPlacementOrDiscard, let pending = state.pendingDraw {
            // Active player drew a card but never placed it: discard the
            // drawn card with no effect and end the turn.
            state.discardPile.append(pending.card)
            state.pendingDraw = nil
        } else if state.phase == .waitingForSpecialResolution {
            // The J/Q/K is already on the discard pile from the prior
            // `discardForEffect` step. We just forfeit the effect.
            state.pendingDraw = nil
        }
        endTurn(now: now)
        return true
    }

    private mutating func startTurnTimer(now: Date = Date()) {
        guard state.winnerID == nil else { return }
        guard state.phase != .initialPeek else { return }
        state.currentTurnEndsAt = now.addingTimeInterval(CaboGameEngine.turnTimeoutSeconds)
    }

    mutating func attemptMatchDiscard(playerID: UUID, handIndex: Int) throws -> Bool {
        guard state.winnerID == nil else { throw GameRuleError.roundFinished }
        guard state.phase != .initialPeek else { throw GameRuleError.invalidPhase }
        state.kingSwapHighlight = nil
        guard let topDiscard = state.discardPile.last else { throw GameRuleError.noDiscardCard }

        let playerIndex = try indexOfPlayer(playerID)
        guard state.players[playerIndex].hand.indices.contains(handIndex) else {
            throw GameRuleError.invalidCardIndex
        }

        guard let selected = state.players[playerIndex].hand[handIndex] else {
            throw GameRuleError.invalidCardIndex
        }
        if selected.rank == topDiscard.rank {
            state.players[playerIndex].hand[handIndex] = nil
            state.discardPile.append(selected)
            return true
        }

        // Failed match: end the matcher's *current* turn if it is theirs;
        // otherwise (an off-turn match attempt) make them sit out their next
        // full turn.
        if state.currentPlayerIndex == playerIndex {
            // Discard any pending drawn card (without applying its effect)
            // so deck/discard counts stay consistent before advancing turn.
            if let pending = state.pendingDraw {
                state.discardPile.append(pending.card)
                state.pendingDraw = nil
            }
            endTurn()
        } else {
            state.players[playerIndex].roundsToSkip += 1
        }
        return false
    }

    mutating func drawCard(for playerID: UUID, source: DrawSource) throws -> Card {
        try validateTurn(for: playerID)
        guard state.phase == .waitingForDraw else { throw GameRuleError.invalidPhase }
        state.kingSwapHighlight = nil

        let card: Card
        switch source {
        case .deck:
            card = try drawFromDeckInternal()
        case .discardTop:
            guard let discardTop = state.discardPile.popLast() else {
                throw GameRuleError.noDiscardCard
            }
            card = discardTop
        }

        state.pendingDraw = PendingDraw(card: card, source: source)
        state.phase = .waitingForPlacementOrDiscard
        return card
    }

    mutating func replaceCard(for playerID: UUID, at handIndex: Int) throws {
        try validateTurn(for: playerID)
        guard state.phase == .waitingForPlacementOrDiscard else { throw GameRuleError.invalidPhase }
        guard let pending = state.pendingDraw else { throw GameRuleError.noPendingDraw }
        state.kingSwapHighlight = nil

        let playerIndex = try indexOfPlayer(playerID)
        guard state.players[playerIndex].hand.indices.contains(handIndex) else {
            throw GameRuleError.invalidCardIndex
        }

        let replaced = state.players[playerIndex].hand[handIndex]
        state.players[playerIndex].hand[handIndex] = pending.card
        if let replaced {
            state.discardPile.append(replaced)
        }

        state.pendingDraw = nil
        endTurn()
    }

    mutating func discardDrawnCardAndUseEffect(for playerID: UUID) throws -> Rank {
        try validateTurn(for: playerID)
        guard state.phase == .waitingForPlacementOrDiscard else { throw GameRuleError.invalidPhase }
        guard let pending = state.pendingDraw else { throw GameRuleError.noPendingDraw }

        state.kingSwapHighlight = nil
        state.discardPile.append(pending.card)
        state.pendingDraw = nil

        switch pending.card.rank {
        case .jack, .queen, .king:
            state.phase = .waitingForSpecialResolution
        default:
            endTurn()
        }
        return pending.card.rank
    }

    mutating func resolveSpecialAction(for playerID: UUID, action: SpecialAction) throws {
        try validateTurn(for: playerID)
        guard state.phase == .waitingForSpecialResolution else { throw GameRuleError.invalidPhase }
        state.kingSwapHighlight = nil

        guard let top = state.discardPile.last else { throw GameRuleError.noDiscardCard }
        switch (top.rank, action) {
        case (.jack, .lookOwnCard(let targetPlayerID, let cardIndex)):
            guard targetPlayerID == playerID else { throw GameRuleError.invalidSpecialCard }
            let playerIndex = try indexOfPlayer(playerID)
            guard state.players[playerIndex].hand.indices.contains(cardIndex) else {
                throw GameRuleError.invalidCardIndex
            }
            guard state.players[playerIndex].hand[cardIndex] != nil else {
                throw GameRuleError.invalidCardIndex
            }

        case (.queen, .lookOtherCard(let targetPlayerID, let cardIndex)):
            let targetIdx = try indexOfPlayer(targetPlayerID)
            guard targetPlayerID != playerID else { throw GameRuleError.invalidSpecialCard }
            guard state.players[targetIdx].hand.indices.contains(cardIndex) else {
                throw GameRuleError.invalidCardIndex
            }
            // Queen allows peeking an opponent card this turn.
            // Client can surface this to the active player; persistent memory is not stored for opponents.
            guard state.players[targetIdx].hand[cardIndex] != nil else {
                throw GameRuleError.invalidCardIndex
            }

        case (.king, .swapCards(let fromPlayerID, let fromCardIndex, let toPlayerID, let toCardIndex)):
            let fromIdx = try indexOfPlayer(fromPlayerID)
            let toIdx = try indexOfPlayer(toPlayerID)
            guard fromPlayerID == playerID, toPlayerID != playerID else {
                throw GameRuleError.invalidSpecialCard
            }
            guard state.players[fromIdx].hand.indices.contains(fromCardIndex),
                  state.players[toIdx].hand.indices.contains(toCardIndex) else {
                throw GameRuleError.invalidCardIndex
            }
            guard state.players[fromIdx].hand[fromCardIndex] != nil,
                  state.players[toIdx].hand[toCardIndex] != nil else {
                throw GameRuleError.invalidCardIndex
            }
            let tmp = state.players[fromIdx].hand[fromCardIndex]
            state.players[fromIdx].hand[fromCardIndex] = state.players[toIdx].hand[toCardIndex]
            state.players[toIdx].hand[toCardIndex] = tmp
            state.kingSwapHighlight = KingSwapHighlight(
                fromPlayerID: fromPlayerID,
                fromHandIndex: fromCardIndex,
                toPlayerID: toPlayerID,
                toHandIndex: toCardIndex
            )

        default:
            throw GameRuleError.invalidSpecialCard
        }

        endTurn()
    }

    mutating func callCabo(for playerID: UUID) throws {
        try validateTurn(for: playerID)
        guard state.phase == .waitingForDraw else { throw GameRuleError.invalidPhase }
        guard state.pendingDraw == nil else { throw GameRuleError.invalidPhase }
        guard state.caboCallerID == nil else { throw GameRuleError.caboAlreadyCalled }

        state.caboCallerID = playerID
        state.kingSwapHighlight = nil
        endTurn()
    }

    private mutating func drawFromDeckInternal() throws -> Card {
        if state.deck.isEmpty {
            try reshuffleDiscardIntoDeckIfNeeded()
        }
        guard let card = state.deck.popLast() else {
            throw GameRuleError.emptyDeck
        }
        return card
    }

    private mutating func reshuffleDiscardIntoDeckIfNeeded() throws {
        guard state.deck.isEmpty else { return }
        guard state.discardPile.count > 1 else { throw GameRuleError.emptyDeck }
        let keepTop = state.discardPile.removeLast()
        state.deck = state.discardPile.shuffled()
        state.discardPile = [keepTop]
    }

    private mutating func endTurn(now: Date = Date()) {
        state.phase = .waitingForDraw
        if !state.players.isEmpty {
            var nextIndex = state.currentPlayerIndex
            var safety = state.players.count
            repeat {
                nextIndex = (nextIndex + 1) % state.players.count
                if state.players[nextIndex].roundsToSkip > 0 {
                    state.players[nextIndex].roundsToSkip -= 1
                } else {
                    break
                }
                safety -= 1
            } while safety > 0
            state.currentPlayerIndex = nextIndex

            if let caboCallerID = state.caboCallerID,
               state.players[nextIndex].id == caboCallerID {
                finishRound()
                return
            }
        }
        startTurnTimer(now: now)
    }

    private mutating func finishRound() {
        let sorted = state.players.sorted { $0.score() < $1.score() }
        state.winnerID = sorted.first?.id
        state.caboCallerID = nil
        state.phase = .waitingForDraw
        state.pendingDraw = nil
        state.currentTurnEndsAt = nil
        state.kingSwapHighlight = nil
    }

    private func indexOfPlayer(_ playerID: UUID) throws -> Int {
        guard let idx = state.players.firstIndex(where: { $0.id == playerID }) else {
            throw GameRuleError.invalidPlayer
        }
        return idx
    }

    private func validateTurn(for playerID: UUID) throws {
        guard state.currentPlayerID == playerID else {
            throw GameRuleError.invalidPlayer
        }
    }
}

File: android/app/src/main/kotlin/com/alex/cabogame/engine/CaboGameEngine.kt
package com.alex.cabogame.engine

import com.alex.cabogame.models.*

class GameRuleError(message: String) : Exception(message)

class CaboGameEngine {
    companion object {
        /** Maximum length of a single player turn after the initial peek phase. */
        const val TURN_TIMEOUT_MILLIS: Long = 20_000L
    }

    var state: GameState = GameState()
        private set

    fun addPlayer(name: String): String {
        val player = Player(name = name)
        state = state.copy(players = state.players + player)
        return player.id
    }

    fun load(newState: GameState) {
        state = newState
    }

    fun startGame() {
        if (state.players.size < 2) throw GameRuleError("Need at least 2 players.")

        var players = state.players.map { it.copy(hand = List(4) { null }, roundsToSkip = 0) }
        val deck = caboDeck().toMutableList()

        for (slot in 0 until 4) {
            players = players.mapIndexed { idx, player ->
                val card = deck.removeLast()
                val hand = player.hand.toMutableList().also { it[slot] = card }
                player.copy(hand = hand)
            }
        }

        val firstDiscard = deck.removeLast()

        state = state.copy(
            players = players,
            deck = deck,
            discardPile = listOf(firstDiscard),
            currentPlayerIndex = 0,
            phase = TurnPhase.INITIAL_PEEK,
            pendingDraw = null,
            winnerID = null,
            caboCallerID = null,
            rematchRequestedByHost = false,
            readyPlayerIDs = emptySet(),
            hasStarted = true,
            playersFinishedInitialPeek = 0,
            initialPeekedIndicesByPlayerIndex = List(players.size) { emptyList() },
            // Start the initial peek grace timer immediately (iOS behavior).
            initialPeekGraceEndsAt = System.currentTimeMillis() + 15_000L,
            currentTurnEndsAt = null
        )
    }

    fun peekInitialCard(playerID: String, handIndex: Int): Card {
        if (state.phase != TurnPhase.INITIAL_PEEK) throw GameRuleError("This action is not valid right now.")

        val playerIndex = indexOfPlayer(playerID)
        val player = state.players[playerIndex]
        if (!player.hand.indices.contains(handIndex)) throw GameRuleError("Selected card index is invalid.")

        val alreadyPeeked = state.initialPeekedIndicesByPlayerIndex
            .getOrNull(playerIndex)
            ?.contains(handIndex) == true
        if (alreadyPeeked) throw GameRuleError("Selected card index is invalid.")

        val card = player.hand[handIndex] ?: throw GameRuleError("Selected card index is invalid.")

        val currentForPlayer = state.initialPeekedIndicesByPlayerIndex.getOrNull(playerIndex) ?: emptyList()
        val updatedForPlayer = currentForPlayer + handIndex

        val newInitialPeeked = state.initialPeekedIndicesByPlayerIndex.toMutableList().also { list ->
            // Be defensive: older/mismatched states might have an empty list.
            while (list.size < state.players.size) list.add(emptyList())
            list[playerIndex] = updatedForPlayer
        }

        var newState = state.copy(initialPeekedIndicesByPlayerIndex = newInitialPeeked)

        // Count completion when a player reaches 2 peeked cards.
        if (updatedForPlayer.size == 2) {
            newState = newState.copy(playersFinishedInitialPeek = newState.playersFinishedInitialPeek + 1)
        }

        state = newState
        return card
    }

    fun beginMainTurnsAfterInitialPeekIfReady(nowMillis: Long = System.currentTimeMillis()) {
        if (state.phase != TurnPhase.INITIAL_PEEK) return
        val graceEndsAt = state.initialPeekGraceEndsAt ?: return
        if (nowMillis < graceEndsAt) return

        state = state.copy(
            phase = TurnPhase.WAITING_FOR_DRAW,
            currentPlayerIndex = 0,
            initialPeekedIndicesByPlayerIndex = List(state.players.size) { emptyList() },
            playersFinishedInitialPeek = 0,
            initialPeekGraceEndsAt = null,
            currentTurnEndsAt = nowMillis + TURN_TIMEOUT_MILLIS
        )
    }

    /**
     * Forces the current player's turn to end because the per-turn timer
     * expired. Behavior depends on the phase the player got stuck in:
     *  - WAITING_FOR_DRAW: skip the turn outright.
     *  - WAITING_FOR_PLACEMENT_OR_DISCARD: discard the pending drawn card without
     *    applying its effect, then advance.
     *  - WAITING_FOR_SPECIAL_RESOLUTION: forfeit the J/Q/K effect (the card is
     *    already on the discard pile), advance.
     * Returns true if a turn was actually ended, false if there was nothing
     * to do (e.g. game over or initial peek phase).
     */
    fun enforceTurnTimeoutIfNeeded(nowMillis: Long = System.currentTimeMillis()): Boolean {
        if (state.winnerID != null) return false
        if (state.phase == TurnPhase.INITIAL_PEEK) return false
        val deadline = state.currentTurnEndsAt ?: return false
        if (nowMillis < deadline) return false

        if (state.phase == TurnPhase.WAITING_FOR_PLACEMENT_OR_DISCARD) {
            val pending = state.pendingDraw
            if (pending != null) {
                state = state.copy(
                    discardPile = state.discardPile + pending.card,
                    pendingDraw = null
                )
            }
        } else if (state.phase == TurnPhase.WAITING_FOR_SPECIAL_RESOLUTION) {
            // The J/Q/K is already on the discard pile from the prior
            // discardForEffect step; we just forfeit the effect.
            state = state.copy(pendingDraw = null)
        }
        endTurn(nowMillis)
        return true
    }

    fun attemptMatchDiscard(playerID: String, handIndex: Int): Boolean {
        if (state.winnerID != null) throw GameRuleError("Round already finished.")
        if (state.phase == TurnPhase.INITIAL_PEEK) throw GameRuleError("This action is not valid right now.")
        val topDiscard = state.discardPile.lastOrNull() ?: throw GameRuleError("No card in discard pile.")

        val playerIndex = indexOfPlayer(playerID)
        val player = state.players[playerIndex]
        if (!player.hand.indices.contains(handIndex)) throw GameRuleError("Selected card index is invalid.")
        val selected = player.hand[handIndex] ?: throw GameRuleError("Selected card index is invalid.")

        return if (selected.rank == topDiscard.rank) {
            val newHand = player.hand.toMutableList().also { it[handIndex] = null }
            val newPlayers = state.players.toMutableList().also {
                it[playerIndex] = player.copy(hand = newHand)
            }
            state = state.copy(players = newPlayers, discardPile = state.discardPile + selected)
            true
        } else {
            // Failed match: end the matcher's *current* turn if it is theirs;
            // otherwise (an off-turn match attempt) make them sit out their next
            // full turn.
            if (state.currentPlayerIndex == playerIndex) {
                // Discard any pending drawn card (without applying its effect)
                // so deck/discard counts stay consistent before advancing turn.
                val pending = state.pendingDraw
                if (pending != null) {
                    state = state.copy(
                        discardPile = state.discardPile + pending.card,
                        pendingDraw = null
                    )
                }
                endTurn()
            } else {
                val newPlayers = state.players.toMutableList().also {
                    it[playerIndex] = player.copy(roundsToSkip = player.roundsToSkip + 1)
                }
                state = state.copy(players = newPlayers)
            }
            false
        }
    }

    fun drawCard(playerID: String, source: DrawSource): Card {
        validateTurn(playerID)
        if (state.phase != TurnPhase.WAITING_FOR_DRAW) throw GameRuleError("This action is not valid right now.")

        val card = when (source) {
            DrawSource.DECK -> drawFromDeckInternal()
            DrawSource.DISCARD_TOP -> {
                state.discardPile.lastOrNull() ?: throw GameRuleError("No card in discard pile.")
                    .also { state = state.copy(discardPile = state.discardPile.dropLast(1)) }
                val top = state.discardPile.last()
                state = state.copy(discardPile = state.discardPile.dropLast(1))
                top
            }
        }

        state = state.copy(
            pendingDraw = PendingDraw(card, source),
            phase = TurnPhase.WAITING_FOR_PLACEMENT_OR_DISCARD
        )
        return card
    }

    fun replaceCard(playerID: String, handIndex: Int) {
        validateTurn(playerID)
        if (state.phase != TurnPhase.WAITING_FOR_PLACEMENT_OR_DISCARD) throw GameRuleError("This action is not valid right now.")
        val pending = state.pendingDraw ?: throw GameRuleError("No drawn card is pending.")

        val playerIndex = indexOfPlayer(playerID)
        val player = state.players[playerIndex]
        if (!player.hand.indices.contains(handIndex)) throw GameRuleError("Selected card index is invalid.")

        val replaced = player.hand[handIndex]
        val newHand = player.hand.toMutableList().also { it[handIndex] = pending.card }
        val newPlayers = state.players.toMutableList().also {
            it[playerIndex] = player.copy(hand = newHand)
        }
        val newDiscard = if (replaced != null) state.discardPile + replaced else state.discardPile

        state = state.copy(players = newPlayers, discardPile = newDiscard, pendingDraw = null)
        endTurn()
    }

    fun discardDrawnCardAndUseEffect(playerID: String): Rank {
        validateTurn(playerID)
        if (state.phase != TurnPhase.WAITING_FOR_PLACEMENT_OR_DISCARD) throw GameRuleError("This action is not valid right now.")
        val pending = state.pendingDraw ?: throw GameRuleError("No drawn card is pending.")

        state = state.copy(discardPile = state.discardPile + pending.card, pendingDraw = null)

        when (pending.card.rank) {
            Rank.JACK, Rank.QUEEN, Rank.KING ->
                state = state.copy(phase = TurnPhase.WAITING_FOR_SPECIAL_RESOLUTION)
            else -> endTurn()
        }
        return pending.card.rank
    }

    fun resolveSpecialAction(playerID: String, action: SpecialAction) {
        validateTurn(playerID)
        if (state.phase != TurnPhase.WAITING_FOR_SPECIAL_RESOLUTION) throw GameRuleError("This action is not valid right now.")
        val top = state.discardPile.lastOrNull() ?: throw GameRuleError("No card in discard pile.")

        when {
            top.rank == Rank.JACK && action is SpecialAction.LookOwnCard -> {
                if (action.playerID != playerID) throw GameRuleError("This card has no special power.")
                val idx = indexOfPlayer(playerID)
                val player = state.players[idx]
                if (!player.hand.indices.contains(action.cardIndex) || player.hand[action.cardIndex] == null)
                    throw GameRuleError("Selected card index is invalid.")
            }
            top.rank == Rank.QUEEN && action is SpecialAction.LookOtherCard -> {
                if (action.targetPlayerID == playerID) throw GameRuleError("This card has no special power.")
                val idx = indexOfPlayer(action.targetPlayerID)
                val player = state.players[idx]
                if (!player.hand.indices.contains(action.cardIndex) || player.hand[action.cardIndex] == null)
                    throw GameRuleError("Selected card index is invalid.")
            }
            top.rank == Rank.KING && action is SpecialAction.SwapCards -> {
                if (action.fromPlayerID != playerID || action.toPlayerID == playerID)
                    throw GameRuleError("This card has no special power.")
                val fromIdx = indexOfPlayer(action.fromPlayerID)
                val toIdx = indexOfPlayer(action.toPlayerID)
                val fromPlayer = state.players[fromIdx]
                val toPlayer = state.players[toIdx]
                if (!fromPlayer.hand.indices.contains(action.fromCardIndex) || fromPlayer.hand[action.fromCardIndex] == null)
                    throw GameRuleError("Selected card index is invalid.")
                if (!toPlayer.hand.indices.contains(action.toCardIndex) || toPlayer.hand[action.toCardIndex] == null)
                    throw GameRuleError("Selected card index is invalid.")

                val newPlayers = state.players.toMutableList()
                val fromHand = fromPlayer.hand.toMutableList()
                val toHand = toPlayer.hand.toMutableList()
                val tmp = fromHand[action.fromCardIndex]
                fromHand[action.fromCardIndex] = toHand[action.toCardIndex]
                toHand[action.toCardIndex] = tmp
                newPlayers[fromIdx] = fromPlayer.copy(hand = fromHand)
                newPlayers[toIdx] = toPlayer.copy(hand = toHand)
                state = state.copy(players = newPlayers)
            }
            else -> throw GameRuleError("This card has no special power.")
        }

        endTurn()
    }

    fun callCabo(playerID: String) {
        validateTurn(playerID)
        if (state.phase != TurnPhase.WAITING_FOR_DRAW) throw GameRuleError("This action is not valid right now.")
        if (state.pendingDraw != null) throw GameRuleError("This action is not valid right now.")
        if (state.caboCallerID != null) throw GameRuleError("Cabo has already been called this round.")
        state = state.copy(caboCallerID = playerID)
        endTurn()
    }

    private fun drawFromDeckInternal(): Card {
        if (state.deck.isEmpty()) reshuffleDiscardIntoDeck()
        if (state.deck.isEmpty()) throw GameRuleError("Deck is empty.")
        val card = state.deck.last()
        state = state.copy(deck = state.deck.dropLast(1))
        return card
    }

    private fun reshuffleDiscardIntoDeck() {
        if (state.discardPile.size <= 1) throw GameRuleError("Deck is empty.")
        val keepTop = state.discardPile.last()
        state = state.copy(deck = state.discardPile.dropLast(1).shuffled(), discardPile = listOf(keepTop))
    }

    private fun endTurn(nowMillis: Long = System.currentTimeMillis()) {
        var next = state.currentPlayerIndex
        var safety = state.players.size
        val players = state.players.toMutableList()

        do {
            next = (next + 1) % players.size
            if (players[next].roundsToSkip > 0) {
                players[next] = players[next].copy(roundsToSkip = players[next].roundsToSkip - 1)
            } else {
                break
            }
            safety--
        } while (safety > 0)

        state = state.copy(players = players, currentPlayerIndex = next, phase = TurnPhase.WAITING_FOR_DRAW)

        if (state.caboCallerID != null && state.players[next].id == state.caboCallerID) {
            finishRound()
            return
        }

        // Reset the turn timer for the new active player.
        state = state.copy(currentTurnEndsAt = nowMillis + TURN_TIMEOUT_MILLIS)
    }

    private fun finishRound() {
        val winner = state.players.minByOrNull { it.score() }
        state = state.copy(
            winnerID = winner?.id,
            caboCallerID = null,
            phase = TurnPhase.WAITING_FOR_DRAW,
            pendingDraw = null,
            currentTurnEndsAt = null
        )
    }

    private fun indexOfPlayer(playerID: String): Int =
        state.players.indexOfFirst { it.id == playerID }
            .also { if (it == -1) throw GameRuleError("It is not this player's turn.") }

    private fun validateTurn(playerID: String) {
        if (state.currentPlayerID != playerID) throw GameRuleError("It is not this player's turn.")
    }
}
